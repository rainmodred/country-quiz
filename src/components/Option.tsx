/** @jsxImportSource @emotion/react */
import { AiOutlineCheckCircle, AiOutlineCloseCircle } from 'react-icons/ai';
import { css } from '@emotion/react';

import { Answer } from 'types';
import { colors } from '../styles/colors';

interface OptionProps {
  answer: Answer;
  wrong?: boolean;
  disabled: boolean;
  onClick: (answer: Answer) => void;
}

export default function Option({
  answer,
  disabled = false,
  wrong = false,
  onClick,
}: OptionProps): JSX.Element {
  const { variant, text, correct } = answer;

  const isWrong = wrong && disabled;
  const isCorrect = correct && disabled;

  function handleClick(
    event: React.MouseEvent<HTMLButtonElement, MouseEvent>,
    currentAnswer: Answer,
  ) {
    event.currentTarget.blur();
    onClick(currentAnswer);
  }

  let bgColor = 'none';
  if (isWrong) {
    bgColor = colors.colorWrongBg;
  }
  if (isCorrect) {
    bgColor = colors.colorCorrectBg;
  }

  return (
    <button
      onClick={event => handleClick(event, answer)}
      type="button"
      css={css`
        border-radius: 12px;
        border: ${isWrong || isCorrect
          ? 'none'
          : `2px solid ${colors.colorOptionBorder}`};
        background: ${bgColor};
        max-height: 56px;
        width: 100%;
        max-width: 400px;
        overflow: hidden;
        transition: all 0.2s ease-in-out;
        color: ${isWrong || isCorrect
          ? colors.colorWhite
          : colors.colorOptionText};
        cursor: pointer;
        display: flex;
        align-items: center;
        padding: 15px 18px 14px 19px;
        pointer-events: ${disabled && 'none'};
        & > span {
          font-size: 18px;
          line-height: 27px;
          font-weight: 500;
        }
        &:hover {
          background-color: ${bgColor === 'none' ? colors.colorHover : bgColor};
          color: ${colors.colorWhite};
          border: none;
        }
      `}
    >
      <span
        css={css`
          margin-right: 45px;
        `}
      >
        {variant}
      </span>
      <span
        css={css`
          margin-right: auto;
          white-space: nowrap;
          overflow: hidden;
        `}
      >
        {text}
      </span>
      {isWrong && <AiOutlineCloseCircle size="20" title="closeIcon" />}
      {isCorrect && <AiOutlineCheckCircle size="20" title="checkIcon" />}
    </button>
  );
}
