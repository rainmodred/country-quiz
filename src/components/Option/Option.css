.option {
  border-radius: 12px;
  border: 2px solid var(--color-option-border);
  background: none;
  width: 400px;
  max-height: 56px;
  transition: all 0.2s ease-in;
  color: var(--color-option-text);
  cursor: pointer;
  display: flex;
  align-items: center;
  padding: 15px 18px 14px 19px;
}

.option:hover {
  background-color: var(--color-hover);
  color: var(--color-white);
  border: none;
}

.option > span {
  font-size: 18px;
  line-height: 27px;
  font-weight: 500;
}

.option__id {
  margin-right: 45px;
}

.option__text {
  margin-right: auto;
  white-space: nowrap;
  overflow: hidden;
}

.option--error {
  background-color: var(--color-error-bg);
  border: none;
  color: var(--color-white);
}

.option--error:hover {
  background-color: currentColor;
}

.option--correct {
  background-color: var(--color-correct-bg);
  border: none;
  color: var(--color-white);
}

.option--disabled {
  pointer-events: none;
}
