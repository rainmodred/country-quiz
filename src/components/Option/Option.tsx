import { AiOutlineCheckCircle, AiOutlineCloseCircle } from 'react-icons/ai';
import './Option.css';

interface OptionProps {
  variant: string;
  text: string;
  disabled: boolean;
  onClick: (text: string) => void;
  error?: boolean;
  correct?: boolean;
}

export default function Option({
  variant,
  text,
  error = false,
  correct = false,
  disabled = false,
  onClick,
}: OptionProps): JSX.Element {
  const isError = error && disabled ? 'option--error' : '';
  const isCorrect = correct && disabled ? 'option--correct' : '';
  const isDisabled = disabled ? 'option--disabled' : '';

  const classes = `option ${isError} ${isCorrect} ${isDisabled}`;

  function handleClick(
    event: React.MouseEvent<HTMLButtonElement, MouseEvent>,
    answer: string,
  ) {
    event.currentTarget.blur();
    onClick(answer);
  }

  return (
    <button
      onClick={event => handleClick(event, text)}
      type="button"
      className={classes}
    >
      <span className="option__id">{variant}</span>
      <span className="option__text">{text}</span>
      {isError && <AiOutlineCloseCircle size="20" title="closeIcon" />}
      {isCorrect && <AiOutlineCheckCircle size="20" title="checkIcon" />}
    </button>
  );
}
