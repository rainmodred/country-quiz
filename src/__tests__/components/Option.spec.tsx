import { render } from '@testing-library/react';
import Option from '../../components/Option';

describe('Option', () => {
  it('renders Option', () => {
    const { container, getByText } = render(
      <Option
        answer={{ variant: 'A', text: 'test', correct: false }}
        onClick={() => {}}
        disabled={false}
      />,
    );

    expect(getByText(/test/i)).toBeTruthy();
    expect(container.firstChild).toMatchSnapshot();
  });
  it('renders correct Option', () => {
    const { container, getByText, queryByTitle } = render(
      <Option
        answer={{ variant: 'A', text: 'test', correct: true }}
        onClick={() => {}}
        disabled
      />,
    );

    expect(getByText(/test/i)).toBeTruthy();
    expect(queryByTitle(/checkIcon/i)).toBeTruthy();
    expect(container.firstChild).toMatchSnapshot();
  });
  it('renders wrong Option', () => {
    const { container, getByText, queryByTitle } = render(
      <Option
        answer={{ variant: 'A', text: 'test', correct: false }}
        onClick={() => {}}
        disabled
        wrong
      />,
    );

    expect(getByText(/test/i)).toBeTruthy();
    expect(queryByTitle(/closeIcon/i)).toBeTruthy();
    expect(container.firstChild).toMatchSnapshot();
  });
});
