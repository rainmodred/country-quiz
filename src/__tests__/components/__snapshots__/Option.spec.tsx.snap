// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Option renders Option 1`] = `
.emotion-0 {
  border-radius: 12px;
  border: 2px solid rgba(96, 102, 208, 0.7);
  background: none;
  max-height: 56px;
  width: 100%;
  max-width: 400px;
  overflow: hidden;
  -webkit-transition: all 0.2s ease-in-out;
  transition: all 0.2s ease-in-out;
  color: rgba(96, 102, 208, 0.8);
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 15px 18px 14px 19px;
}

.emotion-0>span {
  font-size: 18px;
  line-height: 27px;
  font-weight: 500;
}

.emotion-0:hover {
  background-color: rgba(249, 168, 38, 1);
  color: rgba(255, 255, 255, 1);
  border: none;
}

.emotion-1 {
  margin-right: 45px;
}

.emotion-2 {
  margin-right: auto;
  white-space: nowrap;
  overflow: hidden;
}

<button
  class="emotion-0"
  type="button"
>
  <span
    class="emotion-1"
  >
    A
  </span>
  <span
    class="emotion-2"
  >
    test
  </span>
</button>
`;

exports[`Option renders correct Option 1`] = `
.emotion-0 {
  border-radius: 12px;
  border: none;
  background: rgba(96, 191, 136, 1);
  max-height: 56px;
  width: 100%;
  max-width: 400px;
  overflow: hidden;
  -webkit-transition: all 0.2s ease-in-out;
  transition: all 0.2s ease-in-out;
  color: rgba(255, 255, 255, 1);
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 15px 18px 14px 19px;
  pointer-events: none;
}

.emotion-0>span {
  font-size: 18px;
  line-height: 27px;
  font-weight: 500;
}

.emotion-0:hover {
  background-color: rgba(96, 191, 136, 1);
  color: rgba(255, 255, 255, 1);
  border: none;
}

.emotion-1 {
  margin-right: 45px;
}

.emotion-2 {
  margin-right: auto;
  white-space: nowrap;
  overflow: hidden;
}

<button
  class="emotion-0"
  type="button"
>
  <span
    class="emotion-1"
  >
    A
  </span>
  <span
    class="emotion-2"
  >
    test
  </span>
  <svg
    fill="currentColor"
    height="20"
    stroke="currentColor"
    stroke-width="0"
    viewBox="0 0 1024 1024"
    width="20"
    xmlns="http://www.w3.org/2000/svg"
  >
    <title>
      checkIcon
    </title>
    <path
      d="M699 353h-46.9c-10.2 0-19.9 4.9-25.9 13.3L469 584.3l-71.2-98.8c-6-8.3-15.6-13.3-25.9-13.3H325c-6.5 0-10.3 7.4-6.5 12.7l124.6 172.8a31.8 31.8 0 0 0 51.7 0l210.6-292c3.9-5.3.1-12.7-6.4-12.7z"
    />
    <path
      d="M512 64C264.6 64 64 264.6 64 512s200.6 448 448 448 448-200.6 448-448S759.4 64 512 64zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
    />
  </svg>
</button>
`;

exports[`Option renders wrong Option 1`] = `
.emotion-0 {
  border-radius: 12px;
  border: none;
  background: rgba(234, 130, 130, 1);
  max-height: 56px;
  width: 100%;
  max-width: 400px;
  overflow: hidden;
  -webkit-transition: all 0.2s ease-in-out;
  transition: all 0.2s ease-in-out;
  color: rgba(255, 255, 255, 1);
  cursor: pointer;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  padding: 15px 18px 14px 19px;
  pointer-events: none;
}

.emotion-0>span {
  font-size: 18px;
  line-height: 27px;
  font-weight: 500;
}

.emotion-0:hover {
  background-color: rgba(234, 130, 130, 1);
  color: rgba(255, 255, 255, 1);
  border: none;
}

.emotion-1 {
  margin-right: 45px;
}

.emotion-2 {
  margin-right: auto;
  white-space: nowrap;
  overflow: hidden;
}

<button
  class="emotion-0"
  type="button"
>
  <span
    class="emotion-1"
  >
    A
  </span>
  <span
    class="emotion-2"
  >
    test
  </span>
  <svg
    fill="currentColor"
    height="20"
    stroke="currentColor"
    stroke-width="0"
    viewBox="0 0 1024 1024"
    width="20"
    xmlns="http://www.w3.org/2000/svg"
  >
    <title>
      closeIcon
    </title>
    <path
      d="M685.4 354.8c0-4.4-3.6-8-8-8l-66 .3L512 465.6l-99.3-118.4-66.1-.3c-4.4 0-8 3.5-8 8 0 1.9.7 3.7 1.9 5.2l130.1 155L340.5 670a8.32 8.32 0 0 0-1.9 5.2c0 4.4 3.6 8 8 8l66.1-.3L512 564.4l99.3 118.4 66 .3c4.4 0 8-3.5 8-8 0-1.9-.7-3.7-1.9-5.2L553.5 515l130.1-155c1.2-1.4 1.8-3.3 1.8-5.2z"
    />
    <path
      d="M512 65C264.6 65 64 265.6 64 513s200.6 448 448 448 448-200.6 448-448S759.4 65 512 65zm0 820c-205.4 0-372-166.6-372-372s166.6-372 372-372 372 166.6 372 372-166.6 372-372 372z"
    />
  </svg>
</button>
`;
